### PR TITLE
Adding a DQM module for dEdxHit info

### PR DESCRIPTION
**Pull request by quertenmont**
This PR add a DQM module to monitor dEdxHit info.
And also fix a small bug in the dEdxHitInfo producer in standard RECO.
Hope this can be included in the release that will be used for both MC and Data reconstruction
Backport of this changes for 74X are also there

### CODE REVIEW

**Comment by quertenmont** [comment]
@cmsbuild please test

**Comment by quertenmont** [comment]
By the way, I don't understand why Calibration/* and DQMOffline/* modules are also listed in the PR,
I didn't touched them.  I suspect they have been included against my will in the rebase from 74X to 75X, but I am not sure how I should proceed then

**Comment by slava77** [comment]
Loic, why do you have the commits from Sunanda here?
Which IB did you start from? These extra commits better go away if they are unrelated.
Could you please update.

**Comment by quertenmont** [comment]
Yes, as I said I don't understand whyt Sunandra stuff is in this PR.
What I did is a rebase of my 742 tag to 750preX.
And I solved all conflict with a "checkout --theirs"
So I don't understand what is hapenning here.

Is there anyway to tell github to forget about the Sunandra commits ?
Or should I stupidly duplicate in 75X all the changes I did in 74X

Please help.  The FAQ are really hard to understand regarding this simple (and very regular) problem.

**Comment by makortel** [comment]
Something along e.g.
```
# within CMSSW_7_5_X_2015-05-18-1100 area, which seems to be the base of this PR
git checkout -b foo CMSSW_7_5_X_2015-05-18-1100
git cherry-pick d8dd78a
git push -f my-cmssw foo:dEdxHitDQM75X
```
should do the job.

**Comment by quertenmont** [comment]
@makortel   Thanks a lot!
I save this to my bookmarks!

**Comment by diguida** [comment]
@quertenmont most likely they were included because either your branch departed the vanilla 74X one before the merge of the latest modifications by HCAL developers went into it, or you made a merge topic which added these commits in PR 8752.
You can clean up the branch the following:

````
cmsrel CMSSW_7_5_X_2015-05-17-1100 #latest IB you used
git cms-init
git cms-addpkg DQM/TrackingMonitor
git cms-addpkg DQM/TrackingMonitorSource
git cms-addpkg RecoTracker/DeDx
git checkout -b dEdxHitDQM75X_v2 from-CMSSW_7_5_X_2015-05-18-1100
git branch -u official-cmssw/CMSSW_7_5_X #use remote tracking branches
git cherry-pick d8dd78a
git push my-cmssw +dEdxHitDQM75X_v2:dEdxHitDQM75X
````

**Comment by diguida** [comment]
@makortel sorry I have just seen you sent the same recipe! I have added the sparse checkout after initialising the local git repo, and the remote tracking branches for utility. For the rest, they are the same!

File: DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+#include "DataFormats/Common/interface/ValueMap.h"
+#include "DataFormats/TrackReco/interface/Track.h"
+#include "DataFormats/TrackReco/interface/TrackFwd.h"
+#include "DataFormats/TrackReco/interface/DeDxHitInfo.h"
+
+
+
+
+
+class DQMStore;
```
Comment (deguio) [inline DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h]: please remove

File: DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  
+  //  virtual void beginRun(const edm::Run&, const edm::EventSetup&); 
+  virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&);
+  virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&);
+
+  void bookHistograms(DQMStore::IBooker &, edm::Run const &, edm::EventSetup const &) override;
+  
+ private:
+  // ----------member data ---------------------------
+  DQMStore * dqmStore_;
```
Comment (deguio) [inline DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h]: please remove

File: DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  
+  virtual void beginJob();
+  virtual void analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup);
+  virtual void endJob() ;
+
+  double harmonic2(const reco::DeDxHitInfo* dedxHits);
+  
+  //  virtual void beginRun(const edm::Run&, const edm::EventSetup&); 
+  virtual void beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&);
```
Comment (deguio) [inline DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h]: please remove

File: DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  virtual void endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&);
```
Comment (deguio) [inline DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h]: please remove

File: DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+class dEdxHitAnalyzer : public DQMEDAnalyzer {
+ public:
+  explicit dEdxHitAnalyzer(const edm::ParameterSet&);
+  ~dEdxHitAnalyzer();
+  
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  
+  virtual void beginJob();
+  virtual void analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup);
+  virtual void endJob() ;
```
Comment (deguio) [inline DQM/TrackingMonitor/interface/dEdxHitAnalyzer.h]: please remove

File: DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+dEdxHitAnalyzer::~dEdxHitAnalyzer() 
+{ 
+
+  if (genTriggerEventFlag_)      delete genTriggerEventFlag_;
+}
+
+// ------------ method called once each job just after ending the event loop  ------------
+void 
+dEdxHitAnalyzer::endJob()
```
Comment (deguio) [inline DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc]: direct access to the DQMStore is not allowed anymore

File: DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "DQMServices/Core/interface/DQMStore.h"
+
+#include <string>
+#include "TMath.h"
+
+dEdxHitAnalyzer::dEdxHitAnalyzer(const edm::ParameterSet& iConfig) 
+  : dqmStore_( edm::Service<DQMStore>().operator->() )
```
Comment (deguio) [inline DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc]: please remove

File: DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+//---------------------------------------------------------------------------------//
+void dEdxHitAnalyzer::beginRun(const edm::Run& iRun, const edm::EventSetup& iSetup)
+{
+ 
+  // Initialize the GenericTriggerEventFlag
+  if ( genTriggerEventFlag_->on() ) genTriggerEventFlag_->initRun( iRun, iSetup );
+}
+*/
+
+void dEdxHitAnalyzer::bookHistograms(DQMStore::IBooker & ibooker,
```
Comment (deguio) [inline DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc]: please move everything which is not pure booking outside the bookhistograms method. you can use the dqmBeginRun instead.
the getParameter can be done in the ctor instead.

File: DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+
+         histname =  "Harm2_dEdxPerTrack_"; 
+         dEdxMEsVector[i].ME_Harm2DeDx = ibooker.book1D(histname, histname,dEdxHarm2Bin, dEdxHarm2Min, dEdxHarm2Max);
+         dEdxMEsVector[i].ME_Harm2DeDx->setAxisTitle("Harmonic2 dEdx estimator for each Track");
+         dEdxMEsVector[i].ME_Harm2DeDx->setAxisTitle("Number of Tracks", 2);
+       }
+    }
+}
+
+void dEdxHitAnalyzer::beginJob()
```
Comment (deguio) [inline DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc]: please remove

File: DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+            }
+         }
+      }
+   }
+}
+
+ 
+
+void 
+dEdxHitAnalyzer::beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&)
```
Comment (deguio) [inline DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc]: remove

File: DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+{
+}
+
+// ------------ method called when ending the processing of a luminosity block  ------------
+void 
+dEdxHitAnalyzer::endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&)
```
Comment (deguio) [inline DQM/TrackingMonitor/src/dEdxHitAnalyzer.cc]: remove

**Comment by deguio** [comment]
-1
@quertenmont 
please follow the recommendations at
https://twiki.cern.ch/twiki/bin/viewauth/CMS/ThreadedDQM
feel free to ask if you have questions.
rejecting for now.

for info: if you manage to fix this one by thu/fri we have a chance to have it included in 75.

ciao and thanks for this,
F.

**Comment by quertenmont** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/9129	cdac70b
- the reco part, a fix in RecoTracker/DeDx/plugins/DeDxHitInfoProducer.cc, appears correct (and visually the last one of the kind)
- jenkins tests are successful (and no differences shown in 74X version of this PR)

**Comment by slava77** [comment]
-1

I just realized that instead of monitoring a  product created in reco step (dedxHitInfo) , the DeDxHitInfoProducer is cloned and dedxDQMHitInfo is monitored instead.
Why?

**Comment by quertenmont** [comment]
@slava77  this is a replica of the dE/dx Estimator monitoring which proceed exactly like this.
I guess it is because in the dEdx estimator monitoring we monitor pixel only, strip only and combined estimators, while only one of the three is active in RECO.
Here indeed, it does not make much sense to clone, I agree.  I am changing the config right now

Thanks
Loic

**Comment by deguio** [comment]
+1
thanks @quertenmont 
the code looks ok to me.
ciao,
F.

**Comment by quertenmont** [comment]
Looks like the bot is happy.
Can we converge on this ASAP?

**Comment by slava77** [comment]
the bot shows some random differences that I haven't seen before; I'm checking locally if these are reproducible

**Comment by quertenmont** [comment]
if you refer to the DQM directory, I guess it's because we've added new material to the DQM, isn't it?

**Comment by slava77** [comment]
The "random differences" are all in the efficiency-like plots of HLT IsoTkMu triggers in run2 MC. They are reproducible.

ZMM (wf 1330)
wf1330_hlt_isotkmu
or a similar one in relMon in jenkins https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_7_5_X_2015-05-19-2300+9129/9508/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15+MINIAODMCUP15/HLT_Muon_Distributions_HLT_IsoTkMu20.html

These are very likely an effect of recompilation of DQM modules and some underlying issue with probably uninitialized variable.

I'll keep this note as a somewhat out of place reference for the issue.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/9129	147d78a
- bug fix in the reco part looks correct and visually there are no similar issues in RecoTracker/DeDx/plugins/DeDxHitInfoProducer.cc
- jenkins tests complete successfully and show no differences in comparisons with baseline (nothing to compare to in plots in the baseline).
- the added DQM plots from "/DQMData/Run 1/Tracking/Run summary/dEdxHits/dedxHitInfo" for ZMM wflow 1330 are displayed below

harm2_dedxpertrack
numberofdedxhitspertrack
pixel_dedxpercluster
strip_dedxpercluster

**Comment by davidlange6** [comment]
+1

**Pull request merged**